xarray/core/resample.py

```diff
@@ -194,14 +194,33 @@ def _interpolate(self, kind='linear'):
                 .format(self._obj.data.name)
             )
 
-        x = self._obj[self._dim].astype('float')
+        from ..coding.cftimeindex import CFTimeIndex
+        from .utils import datetime_to_numeric
+        import cftime as cf
+        import numpy as np
+        if isinstance(self._obj.indexes[self._dim], CFTimeIndex):
+            t = self._obj[self._dim]
+            x = datetime_to_numeric(t, datetime_unit='s')
+            x = x.round()
```

Comment:
One other minor thing -- I'm inclined not to round here.  I realize that in the climate community, with non-standard calendars, it would be pretty unusual to encounter dates with a time resolution finer than one second.  However in xarray my sense is we should avoid making assumptions about this.

Fortunately with recent updates in cftime, this is becoming less and less of an issue.  See the difference in behavior here between cftime 1.0.0 and cftime 1.0.3.4 (the latest version):

#### cftime 1.0.0

Here differences in datetime arithmetic precision lead to small differences in the interpolation results between data indexed by a CFTimeIndex and data indexed by a DatetimeIndex.

```
In [1]: import numpy as np
i
In [2]: import xarray as xr

In [3]: times = xr.cftime_range('1892-01-01T12:00:00', periods=15, freq='5256113T')

In [4]: da_cftime = xr.DataArray(range(15), [('time', times)])

In [5]: da_pandas = xr.DataArray(range(15), [('time', times.to_datetimeindex())])

In [6]: new_times = xr.cftime_range('1892', periods=10, freq='6AS-JUN')

In [7]: da_cftime_interp = da_cftime.interp(time=new_times)

In [8]: da_pandas_interp = da_pandas.interp(time=new_times.to_datetimeindex())

In [9]: np.testing.assert_array_equal(da_cftime_interp.values, da_pandas_interp.values)
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-9-9bc960239b46> in <module>
----> 1 np.testing.assert_array_equal(da_cftime_interp.values, da_pandas_interp.values)

//anaconda/envs/xarray-dev-37/lib/python3.6/site-packages/numpy/testing/_private/utils.py in assert_array_equal(x, y, err_msg, verbose)
    863     __tracebackhide__ = True  # Hide traceback for py.test
    864     assert_array_compare(operator.__eq__, x, y, err_msg=err_msg,
--> 865                          verbose=verbose, header='Arrays are not equal')
    866
    867

//anaconda/envs/xarray-dev-37/lib/python3.6/site-packages/numpy/testing/_private/utils.py in assert_array_compare(comparison, x, y, err_msg, verbose, header, precision, equal_nan, equal_inf)
    787                                 verbose=verbose, header=header,
    788                                 names=('x', 'y'), precision=precision)
--> 789             raise AssertionError(msg)
    790     except ValueError:
    791         import traceback

AssertionError:
Arrays are not equal

(mismatch 60.0%)
 x: array([0.041506, 0.641767, 1.242028, 1.842289, 2.442824, 3.043085,
       3.64362 , 4.243881, 4.844416, 5.444677])
 y: array([0.041506, 0.641767, 1.242028, 1.842289, 2.442824, 3.043085,
       3.64362 , 4.243881, 4.844416, 5.444677])
```
Note that we can use `assert_allclose` with the default tolerances here:
```
In [10]: np.testing.assert_allclose(da_cftime_interp.values, da_pandas_interp.values)
```
#### cftime 1.0.3.4

For this example at least, the issues with datetime arithmetic precision have been resolved.

```
In [1]: import numpy as np

In [2]: import xarray as xr

In [3]: times = xr.cftime_range('1892-01-01T12:00:00', periods=15, freq='5256113T')

In [4]: da_cftime = xr.DataArray(range(15), [('time', times)])

In [5]: da_pandas = xr.DataArray(range(15), [('time', times.to_datetimeindex())])

In [6]: new_times = xr.cftime_range('1892', periods=10, freq='6AS-JUN')

In [7]: da_cftime_interp = da_cftime.interp(time=new_times)

In [8]: da_pandas_interp = da_pandas.interp(time=new_times.to_datetimeindex())

In [9]: np.testing.assert_array_equal(da_cftime_interp.values, da_pandas_interp.values)
```

Any errors in interpolation due to microsecond noise in the ordinal coordinate would be expected to be small (I think we could account for this in the tests by using [`xarray.testing.assert_allclose`](http://xarray.pydata.org/en/stable/generated/xarray.testing.assert_allclose.html), which checks that arrays are equal up to a specified tolerance).